File: runtimes/android/app/src/main/kotlin/dev/flux/app/transport/FluxTransport.kt
package dev.flux.app.transport

import java.io.Closeable

/**
 * The bidirectional wire transport the runtime receives frames over.
 *
 * Mirrors the Swift host's `FluxTransport` protocol (FLUX-006): the runtime
 * subscribes to decoded frame bytes and pushes dispatch events back to the dev
 * server. The real implementation ([OkHttpTransport]) uses OkHttp over
 * WebSocket; tests use [MockTransport] with an in-memory channel (real sockets
 * are exercised in FLUX-023).
 */
public interface FluxTransport : Closeable {
    /** Starts the connection, invoking [onFrame] for each received frame. */
    public fun connect(onFrame: (ByteArray) -> Unit)

    /** Sends a raw dispatch message (tap/event) back to the server. */
    public fun send(bytes: ByteArray)

    /** True once [connect] has completed. */
    public fun isConnected(): Boolean
}

/**
 * An in-memory transport for unit tests: frames are injected via [deliver] and
 * dispatched calls captured in [sent]. No network, no real socket — the
 * contract required for deterministic runtime tests (FLUX-007 acceptance).
 */
public class MockTransport : FluxTransport {
    private val frameSink: MutableList<(ByteArray) -> Unit> = mutableListOf()
    private var connected = false

    /** Frames that were sent back to the (pretend) server via [send]. */
    public val sent: MutableList<ByteArray> = mutableListOf()

    public fun addFrameListener(listener: (ByteArray) -> Unit) {
        frameSink.add(listener)
    }

    /** Injects a raw frame into the transport as if received from the server. */
    public fun deliver(frame: ByteArray) {
        frameSink.toList().forEach { it(frame) }
    }

    override fun connect(onFrame: (ByteArray) -> Unit) {
        frameSink.add(onFrame)
        connected = true
    }

    override fun send(bytes: ByteArray) {
        sent.add(bytes)
    }

    override fun isConnected(): Boolean = connected

    override fun close() {
        frameSink.clear()
        connected = false
    }
}

File: runtimes/ios/Sources/Host/FluxTransport.swift
//  FluxTransport.swift
//  iOS host connection model (FR-017 reconnect UX).
//
//  Defines the transport abstraction the host depends on, plus a small
//  observable connection-state machine that drives the "Reconnecting…" banner.
//  The concrete socket lives in `FluxWebSocketTransport`; tests inject a
//  `MockTransport` so the state machine is exercised without a live socket
//  (mirrors Android's `MockTransport` unit-test contract; real sockets land in
//  FLUX-023).

import Foundation
import FluxHost

/// The connection state surfaced to the UI (FR-017).
public enum ConnectionStatus: Equatable, Sendable {
    /// Socket opening; nothing received yet.
    case connecting
    /// Open and receiving frames.
    case connected
    /// Socket dropped; a retry is scheduled every second.
    case reconnecting
}

/// The bidirectional wire transport the runtime receives frames over.
///
/// Mirrors the Android `FluxTransport` interface (FLUX-007): the host subscribes
/// to raw frame bytes and may push dispatch events back. The concrete
/// implementation (`FluxWebSocketTransport`) uses `URLSessionWebSocketTask`.
@MainActor
public protocol FluxTransport: AnyObject {
    /// The latest connection status (drives the reconnect banner).
    var status: ConnectionStatus { get }

    /// Invoked on the main actor for every received message.
    var onFrame: (@MainActor (Data) -> Void)? { get set }

    /// Invoked on the main actor whenever `status` changes.
    var onStatusChange: (@MainActor (ConnectionStatus) -> Void)? { get set }

    /// Opens the connection and begins delivering frames.
    func connect()

    /// Sends a raw dispatch message (tap/event) back to the server.
    func send(_ bytes: Data)

    /// Closes the connection and cancels any pending retry.
    func close()
}

/// Owns the connection-status state for the host UI.
///
/// A `@MainActor` observable value the SwiftUI layer reads to show/hide the
/// "Reconnecting…" banner. It is driven entirely by transport status changes,
/// so it is unit-testable with an injected `MockTransport` (no socket needed).
@MainActor
public final class HostConnectionState: ObservableObject {
    /// The current status; `true` when the banner should be visible.
    @Published public private(set) var status: ConnectionStatus = .connecting

    /// Whether the reconnect banner is currently shown (FR-017).
    public var isReconnecting: Bool { status == .reconnecting }

    /// Binds the state to a transport: every status change is mirrored here and
    /// `onStatusChange` is forwarded, all on the main actor.
    public func bind(_ transport: FluxTransport) {
        status = transport.status
        transport.onStatusChange = { [weak self] newStatus in
            self?.status = newStatus
        }
    }

    /// Forces a status (used by tests to drive the banner without a socket).
    public func setStatus(_ status: ConnectionStatus) {
        self.status = status
    }
}
